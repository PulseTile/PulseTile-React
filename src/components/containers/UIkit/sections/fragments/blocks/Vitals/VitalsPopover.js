import React, { PureComponent } from 'react';
import _ from 'lodash/fp';
import rangeVital from '../../../../../../../assets/images/range-vital.jpg';
import { hasClass } from '../../../../../../../utils/plugin-helpers.utils';

const POPOVER_WIDTH = 266;

/**
 * This component returns popover for Vitals section
 */
export default class VitalsPopover extends PureComponent {

  state = {
    placement: '',
  };

  componentWillMount() {
    window.addEventListener('resize', () => {
      if (!_.isEmpty(this.popover)) {
        this.popover.classList.remove('in');
      }
    });
    document.addEventListener('click', this.handleClick, false);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.handleClick, false);
  }

  handleClick = (event) => {
    if (!_.isEmpty(this.popover)) {
      const currentPopoverWrap = event.target.closest('.popover-wrap');
      let isOpenPopover = false;
      let currentPopover;
      if (currentPopoverWrap) {
        currentPopover = currentPopoverWrap.querySelector('.popover');
        isOpenPopover = hasClass(currentPopover, 'in')
      }
      this.popover.classList.remove('in');
      if (isOpenPopover) {
        currentPopover.classList.add('in');
      }
    }
  };

  togglePopover = () => {
    let { placement } = this.state;
    const { id } = this.props;
    const popover = this.popover;
    if (!_.isEmpty(this.popover)) {
      const popoverWrap = document.getElementById(`popover-wrap-${id}`);
      const pageWidth = window.innerWidth;
      const offsetPopoverWrapLeft = popoverWrap.getBoundingClientRect().left;
      const freePlaceRight = pageWidth - (offsetPopoverWrapLeft + popover.offsetWidth);
      const freePlaceLeft = offsetPopoverWrapLeft;
      if (freePlaceRight > POPOVER_WIDTH) {
        placement = 'right';
      } else if (freePlaceLeft > POPOVER_WIDTH) {
        placement = 'left';
      } else {
        placement = 'top';
      }
      popover.classList.remove('right', 'left', 'top');
      popover.classList.add(placement);
      popover.classList.toggle('in');
      }
  };

  render() {
    const { title, popoverLabels, vitalStatusesType, detailValue, vitalsAddon, id } = this.props;
    return (
      <div className={`input-group vitals-holder popover-wrap ${vitalStatusesType}`} id={`popover-wrap-${id}`}>
        <div className="form-control input-sm">{detailValue}</div>
        <span className='vitals-addon popover-toggle' onClick={this.togglePopover}>{vitalsAddon}</span>
        {!_.isEmpty(popoverLabels) ? <div className="popover vitals-popover fade" ref={(el) => { this.popover = el; }}>
        <div className="arrow" />
          <div className="popover-inner">
            <h3 className="popover-title">{title}</h3>
            <div className="popover-content">
              <div className="range-vital-labels">
                {popoverLabels.map((label, key) => {
                  return (
                    <div key={key} className={`range-vital-label place-${label.place}`}>{label.text}</div>
                  );
                })}
                <img src={rangeVital} alt="" />
              </div>
            </div>
          </div>
        </div> : null
        }
      </div>
    );
  }
}