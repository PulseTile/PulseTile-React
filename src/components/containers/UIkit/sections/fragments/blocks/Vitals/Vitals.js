import React from 'react';
import VitalsPopover from './VitalsPopover';

const respirationRateLabels = [
  { place: 1, text: '≤ 8' },
  { place: 3, text: '9-11' },
  { place: 4, text: '12-20' },
  { place: 5, text: '21-24' },
  { place: 7, text: '≥ 25' },
];

const heartRateLabels = [
  { place: 1, text: '≤ 40' },
  { place: 3, text: '41-50' },
  { place: 4, text: '51-90' },
  { place: 5, text: '91-110' },
  { place: 6, text: '111-130' },
  { place: 7, text: '≥ 131' },
];

/**
 * This component returns content of Vitals section in Blocks
 */
const Vitals = () => {
  return (
    <div id="vitals-popover" className="ui-section">
      <strong className="ui-title">Vitals - News Input and Popover </strong>
      <div className="ui-sub-section">
        <div className="vitals-group-wrapper">
          <div className="row">
            <div className="col-xs-12 col-sm-6">
              <div className="vitals-group highlighter-wrapper">
                <span className="highlighter-warning"></span>
                <label className="vitals-label">Respiration Rate</label>
                <VitalsPopover
                  title='Respiration Rate'
                  popoverLabels={respirationRateLabels}
                  vitalStatusesType='warning'
                  detailValue='23'
                  vitalsAddon='resps/min'
                  id='respirationRate'
                />
              </div>
            </div>
            <div className="col-xs-12 col-sm-6">
              <div className="vitals-group highlighter-wrapper">
                <span className="highlighter-success"></span>
                <label className="vitals-label">Heart Rate</label>
                <VitalsPopover
                  title='Heart Rate'
                  popoverLabels={heartRateLabels}
                  vitalStatusesType='success'
                  detailValue='45'
                  vitalsAddon='bpm'
                  id='heartRate'
                />
              </div>
            </div>
            <div className="col-xs-12 col-sm-6">
              <div className="vitals-group highlighter-wrapper">
                <span className="highlighter-danger"></span>
                <label className="vitals-label">NEWS Score</label>
                <div className="input-holder vitals-holder danger">
                  <div className="form-control input-sm">12</div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Vitals;

