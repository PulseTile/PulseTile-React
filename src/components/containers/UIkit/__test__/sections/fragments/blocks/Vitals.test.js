import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import Vitals from '../../../../sections/fragments/blocks/Vitals/Vitals';
import VitalsPopover from '../../../../sections/fragments/blocks/Vitals/VitalsPopover';

Enzyme.configure({ adapter: new Adapter() });

describe('Component <Vitals />', () => {
  it('should renders Vitals with props correctly', () => {
    const component = shallow(<Vitals />);
    expect(component).toMatchSnapshot();
  });
});

const heartRateLabels = [
    { place: 1, text: '≤ 40' },
    { place: 3, text: '41-50' },
    { place: 4, text: '51-90' },
    { place: 5, text: '91-110' },
    { place: 6, text: '111-130' },
    { place: 7, text: '≥ 131' },
];

describe('Component <VitalsPopover />', () => {
  it('should renders VitalsPopover with props correctly', () => {
    const component = shallow(<VitalsPopover
      title='Heart Rate'
      popoverLabels={heartRateLabels}
      vitalStatusesType='success'
      detailValue='45'
      vitalsAddon='bpm'
      id='heartRate'
    />);
    component.instance().togglePopover();
    expect(component).toMatchSnapshot();
  });
});
