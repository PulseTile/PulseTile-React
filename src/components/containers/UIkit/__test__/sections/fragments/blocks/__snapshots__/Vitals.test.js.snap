// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component <Vitals /> should renders Vitals with props correctly 1`] = `
<div
  className="ui-section"
  id="vitals-popover"
>
  <strong
    className="ui-title"
  >
    Vitals - News Input and Popover 
  </strong>
  <div
    className="ui-sub-section"
  >
    <div
      className="vitals-group-wrapper"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-12 col-sm-6"
        >
          <div
            className="vitals-group highlighter-wrapper"
          >
            <span
              className="highlighter-warning"
            />
            <label
              className="vitals-label"
            >
              Respiration Rate
            </label>
            <VitalsPopover
              detailValue="23"
              id="respirationRate"
              popoverLabels={
                Array [
                  Object {
                    "place": 1,
                    "text": "≤ 8",
                  },
                  Object {
                    "place": 3,
                    "text": "9-11",
                  },
                  Object {
                    "place": 4,
                    "text": "12-20",
                  },
                  Object {
                    "place": 5,
                    "text": "21-24",
                  },
                  Object {
                    "place": 7,
                    "text": "≥ 25",
                  },
                ]
              }
              title="Respiration Rate"
              vitalStatusesType="warning"
              vitalsAddon="resps/min"
            />
          </div>
        </div>
        <div
          className="col-xs-12 col-sm-6"
        >
          <div
            className="vitals-group highlighter-wrapper"
          >
            <span
              className="highlighter-success"
            />
            <label
              className="vitals-label"
            >
              Heart Rate
            </label>
            <VitalsPopover
              detailValue="45"
              id="heartRate"
              popoverLabels={
                Array [
                  Object {
                    "place": 1,
                    "text": "≤ 40",
                  },
                  Object {
                    "place": 3,
                    "text": "41-50",
                  },
                  Object {
                    "place": 4,
                    "text": "51-90",
                  },
                  Object {
                    "place": 5,
                    "text": "91-110",
                  },
                  Object {
                    "place": 6,
                    "text": "111-130",
                  },
                  Object {
                    "place": 7,
                    "text": "≥ 131",
                  },
                ]
              }
              title="Heart Rate"
              vitalStatusesType="success"
              vitalsAddon="bpm"
            />
          </div>
        </div>
        <div
          className="col-xs-12 col-sm-6"
        >
          <div
            className="vitals-group highlighter-wrapper"
          >
            <span
              className="highlighter-danger"
            />
            <label
              className="vitals-label"
            >
              NEWS Score
            </label>
            <div
              className="input-holder vitals-holder danger"
            >
              <div
                className="form-control input-sm"
              >
                12
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Component <VitalsPopover /> should renders VitalsPopover with props correctly 1`] = `
<div
  className="input-group vitals-holder popover-wrap success"
  id="popover-wrap-heartRate"
>
  <div
    className="form-control input-sm"
  >
    45
  </div>
  <span
    className="vitals-addon popover-toggle"
    onClick={[Function]}
  >
    bpm
  </span>
  <div
    className="popover vitals-popover fade"
  >
    <div
      className="arrow"
    />
    <div
      className="popover-inner"
    >
      <h3
        className="popover-title"
      >
        Heart Rate
      </h3>
      <div
        className="popover-content"
      >
        <div
          className="range-vital-labels"
        >
          <div
            className="range-vital-label place-1"
            key="0"
          >
            ≤ 40
          </div>
          <div
            className="range-vital-label place-3"
            key="1"
          >
            41-50
          </div>
          <div
            className="range-vital-label place-4"
            key="2"
          >
            51-90
          </div>
          <div
            className="range-vital-label place-5"
            key="3"
          >
            91-110
          </div>
          <div
            className="range-vital-label place-6"
            key="4"
          >
            111-130
          </div>
          <div
            className="range-vital-label place-7"
            key="5"
          >
            ≥ 131
          </div>
          <img
            alt=""
            src="range-vital.jpg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
